### PR TITLE
simple _cat kernel, dim 0

### PR DESCRIPTION
**Pull request by Janicewei**

### CODE REVIEW

File: hammerblade/torch/riscv.mk
Comment (yodada) [inline hammerblade/torch/riscv.mk]: So here is the merge conflict -- I think you reverted `fb4a35c34ba6c`

You would need to revert the revert commit ;)

File: hammerblade/torch/tests/test_fallback.py
Comment (yodada) [inline hammerblade/torch/tests/test_fallback.py]: Actually, can you delete this test case? It is meant to test the `fallback` mechanism and make sure it does not do crazy things when not enabled.

File: hammerblade/torch/kernel/kernel_cat.cpp
Comment (yodada) [inline hammerblade/torch/kernel/kernel_cat.cpp:50]: Hmmm, this is a little scary -- basically you have every tile do the same work. If you are not ready to parallelize the work, you should put `if (__bsg_id == 0)`, so tile 0 will do the work while others staying idle

File: aten/src/ATen/native/hammerblade/Cat.cpp
Comment (yodada) [inline aten/src/ATen/native/hammerblade/Cat.cpp]: This piece of code looks strange. `result` is just allocated above `Tensor result = at::empty({0}, tensors[0].options());`, so it is not possible for it to alias with anybody.

Comment (Janicewei) [reply aten/src/ATen/native/hammerblade/Cat.cpp]: updated Cat.cpp. I implemented host code based on functions in TensorShape.cpp

File: hammerblade/torch/tests/test_cat.py
Comment (yodada) [inline hammerblade/torch/tests/test_cat.py:47]: More tests will make this even better ;)

For instance, I think you would want to test the case where you cat tensors with different length and empty tensors. Say, [0,1], [0,1,2,3,4,5], [] -> [0,1,0,1,2,3,4,5]

File: hammerblade/torch/kernel/kernel_cat.cpp
Comment (yodada) [inline hammerblade/torch/kernel/kernel_cat.cpp]: Yeah here is an assumption that isn't always true.

But as a first step, it is ok to fail on the case where input tensors have different length. However, it cannot fail silently -- you would want to check for it and throw an error

File: c10/hammerblade/CMakeLists.txt
Comment (yodada) [inline c10/hammerblade/CMakeLists.txt]: Emmm this file shouldn't be touched. I'll help you revert the changes. Dont worry about it

File: hammerblade/torch/kernel/kernel_cat.cpp
Comment (yodada) [inline hammerblade/torch/kernel/kernel_cat.cpp]: this definitely is valid code, but personally I would avoid this. Doing `++j` inside indexing makes code hard to parse.

**Review by yodada — APPROVED** [review]
ok! Looks good! Let me fix the makefile before we merge.

File: hammerblade/torch/kernel/kernel_cat.cpp
Comment (Janicewei) [reply hammerblade/torch/kernel/kernel_cat.cpp]: Okay! changed

**Comment by yodada** [comment]
Reverted `4f22ea6` -- now we are consistent with master branch. This PR is ready to merge once all checks are done.

@Janicewei have you tried on cosim? ideally we should give it a try before we merge

**Comment by Janicewei** [comment]
> Reverted `4f22ea6` -- now we are consistent with master branch. This PR is ready to merge once all checks are done.
> 
> @Janicewei have you tried on cosim? ideally we should give it a try before we merge

give me a few minutes

**Comment by Janicewei** [comment]
> Reverted `4f22ea6` -- now we are consistent with master branch. This PR is ready to merge once all checks are done.
> 
> @Janicewei have you tried on cosim? ideally we should give it a try before we merge

I ran `pycosim -m pytest hb-pytorch/hammerblade/torch/tests/test_cat.py -vs` and tests passed

**Comment by yodada** [comment]
> > Reverted `4f22ea6` -- now we are consistent with master branch. This PR is ready to merge once all checks are done.
> > @Janicewei have you tried on cosim? ideally we should give it a try before we merge
> 
> I ran `pycosim -m pytest hb-pytorch/hammerblade/torch/tests/test_cat.py -vs` and tests passed

perfect!

**Comment by yodada** [comment]
This PR partially implements `torch._cat` -- currently we only support `_cat` with `dim=0`

**Pull request merged**
